chain/bitcoind_conn.go

```diff
@@ -0,0 +1,273 @@
+		HTTPPostMode:         true,
+	}
+
+	client, err := rpcclient.New(clientCfg, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	conn := &BitcoindConn{
+		chainParams:     chainParams,
+		client:          client,
+		zmqHost:         zmqHost,
+		zmqPollInterval: zmqPollInterval,
+		rescanClients:   make(map[uint64]*BitcoindClient),
+		quit:            make(chan struct{}),
+	}
+
+	return conn, nil
+}
+
+// Start attempts to establish a RPC and ZMQ connection to a bitcoind node. If
+// successful, a goroutine is spawned to read events from the ZMQ connection.
+// It's possible for this function to fail due to a limited number of connection
+// attempts. This is done to prevent waiting forever on the connection to be
+// established in the case that the node is down.
+func (c *BitcoindConn) Start() error {
+	if !atomic.CompareAndSwapInt32(&c.started, 0, 1) {
+		return nil
+	}
+
+	// Verify that the node is running on the expected network.
+	net, err := c.getCurrentNet()
+	if err != nil {
+		c.client.Disconnect()
+		return err
+	}
+	if net != c.chainParams.Net {
+		c.client.Disconnect()
+		return fmt.Errorf("expected network %v, got %v",
+			c.chainParams.Net, net)
+	}
+
+	// Establish a ZMQ connection to retrieve block and transaction
+	// notifications from bitcoind.
+	events := []string{"rawblock", "rawtx"}
+	zmqConn, err := gozmq.Subscribe(
+		c.zmqHost, events, c.zmqPollInterval,
+	)
+	if err != nil {
+		c.client.Disconnect()
+		return err
+	}
+
+	c.wg.Add(1)
+	go c.socketHandler(zmqConn)
+
+	return nil
+}
+
+// Stop terminates the RPC and ZMQ connection to a bitcoind node and removes any
+// active rescan clients.
+func (c *BitcoindConn) Stop() {
+	if !atomic.CompareAndSwapInt32(&c.stopped, 0, 1) {
+		return
+	}
+
+	for _, client := range c.rescanClients {
+		client.Stop()
+	}
+
+	close(c.quit)
+	c.client.Shutdown()
+
+	c.client.WaitForShutdown()
+	c.wg.Wait()
+}
+
+// socketHandler reads events from the ZMQ socket and forwards them along to the
+// current rescan clients.
+//
+// NOTE: This must be run as a goroutine.
+func (c *BitcoindConn) socketHandler(zmqConn *gozmq.Conn) {
+	defer c.wg.Done()
+	defer zmqConn.Close()
+
+	log.Info("Started listening for bitcoind notifications via ZMQ on",
+		c.zmqHost)
+
+out:
+	for {
+		// Before attempting to read from the ZMQ socket, we'll make
+		// sure to check if we've been requested to shut down.
+		select {
+		case <-c.quit:
+			return
+		default:
+		}
+
+		// Poll an event from the ZMQ socket. It's possible that the
+		// connection to the socket continuously times out, so we'll
+		// prevent logging this error to prevent spamming the logs.
+		msgBytes, err := zmqConn.Receive()
+		if err != nil {
+			err, ok := err.(net.Error)
+			if !ok || (ok && !err.Timeout()) {
+				log.Error(err)
+			}
+
+			continue out
+		}
+
+		// We have an event! We'll now determine which kind of event was
+		// read, deserialize it, and report it to the different rescan
+		// clients.
+		switch string(msgBytes[0]) {
```

Comment:
Should have a default case for logging purposes.

Reply:
Fixed.